processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeDeriver.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.apm.processor.zipkin;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.hawkular.apm.api.model.Property;
+import org.hawkular.apm.api.model.events.CompletionTime;
+import org.hawkular.apm.server.api.model.zipkin.Annotation;
+import org.hawkular.apm.server.api.model.zipkin.Span;
+import org.hawkular.apm.server.api.services.SpanCache;
+import org.hawkular.apm.server.api.task.AbstractProcessor;
+import org.hawkular.apm.server.api.task.RetryAttemptException;
+import org.jboss.logging.Logger;
+
+/**
+ * Deriver for completion time of trace reported by zipkin instrumentation.
+ *
+ * @author Pavol Loffay
+ */
+@Singleton
+public class CompletionTimeDeriver extends AbstractProcessor<CompletionTimeProcessing, CompletionTime>{
+
+    private static final Logger log = Logger.getLogger(CompletionTimeDeriver.class.getName());
+
+    private SpanCache spanCache;
+
+
+    public CompletionTimeDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTime processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        /**
+         * TODO HWKAMP-348
+         *
+         * Wait one retry delay before calculating trace completion time. This is not a perfect solution but
+         * there is no information in parent span whether there are any children. Solution might be calculating
+         * trace completion time for all spans (not only for root as it is now) and update duration values directly
+         * in database.
+         */
+        if (completionTimeProcessing.getNumberOfRetries() < 1) {
+            completionTimeProcessing.setNumberOfRetries(completionTimeProcessing.getNumberOfRetries() + 1);
+            return null;
+        }
```

Comment:
The most recent 'end time' represents the timestamp of the `lastAnnotation`. (Instrumentation can leave duration as null) 

your described solution would not work for async scenarions imagine:

Span1 sr[1]ss[2]
Span2 cs[5]cr[20]
Span3 sr[6]ss[19]

Server first receives Span1 and it  has no information about any children spans. It has no context information which tells for how long it should wait for children spans before calculation completion time. It can wait for some defined amout of time. 

I will update this PR with strategy that processor will wait if there was an update in trace (new span)